### ISSUE TITLE
Jobs that fail to transfer output still marked successful

### ISSUE THREAD

**Issue by klannon** [matz-e/lobster#260]
Unfortunately, it looks like our SE went belly up about a day into my scale MC production test.  I didn't notice because even though jobs couldn't transfer their output, they were still being marked as successful.  Here's the relevant section of the log file:

```
>>> executing
/cvmfs/oasis.opensciencegrid.org/mis/osg-wn-client/3.2/3.2.33/el6-x86_64/usr/bin/lcg-cp -b -v -D srmv2 file:///var/condor/execute/dir_52221/worker-133909-52231/t.24934/HIG-RunIISummer15GS-01168.root srm://ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop/store/user/lannon/mcprod_ttW_v1/gs_step/HIG-RunIISummer15GS-01168_24934.root

>>> result is
Using grid catalog type: UNKNOWN
Using grid catalog : (null)
VO name: cms
Checksum type: None
Destination SE type: SRMv2
[SE][Ls][] httpg://ndcms.crc.nd.edu:8443/srm/v2/server: CGSI-gSOAP running on q16copt118.crc.nd.edu reports Error reading token data header: Connection closed
[SE][PrepareToPut][] httpg://ndcms.crc.nd.edu:8443/srm/v2/server: CGSI-gSOAP running on q16copt118.crc.nd.edu reports Error reading token data header: Connection closed
lcg_cp: Communication error on send

Traceback (most recent call last):
  File "task.py", line 146, in check_execution
    yield
  File "task.py", line 790, in <module>
    copy_outputs(data, config, env)
  File "task.py", line 461, in copy_outputs
    raise RuntimeError("no stage-out method succeeded")
RuntimeError: no stage-out method succeeded


>>> executing
timeout 300 xrdfs ndcms.crc.nd.edu stat /store/user/lannon/mcprod_ttW_v1/gs_step/HIG-RunIISummer15GS-01168_24934.root

>>> result is
[ERROR] Server responded with an error: [3011] Unable to locate /store/user/lannon/mcprod_ttW_v1/gs_step/HIG-RunIISummer15GS-01168_24934.root; no such file or directory


Execution time 12277
Exiting with code 210
Reporting ExeExitCode 0
Reporting StageOutExitCode 0
```

So, I think the problem is two-fold:  It looks like the failure to stage out creates an unhandled exception in `task.py`  Furthermore, although the task exits with non-zero exit code, the exe and stageout exit codes are set to zero and the job is reported as successful.  Thoughts?

**Comment by matz-e** [comment]
That's weird. Definitely need to look into that. Since the overall exit code is 210 (stage-out error), the job should be marked as failed, I'm a bit mystified why that doesn't happen.

**Comment by klannon** [comment]
I spent a little time trying to look at this, but I've lost my way in this code since the reorganization and I'm still working on finding my way around again.  If you could point me at the relevant files, it would certainly speed up my attempts to understand and fix this.  I'm hesitant to restart processing again until this is fixed because I have so many bad jobs mixed into good ones that I think I'll just have to start over, which is a bit of a shame.

**Comment by matz-e** [comment]
Task processing happens [here](https://github.com/matz-e/lobster/blob/master/lobster/core/task.py#L169) now. I'm curious why the python wrapper can exit with 210, and this does not get propagated… we could also pull the task exit status out of the JSON parameters, [where we get the CMSSW exit code from](https://github.com/matz-e/lobster/blob/master/lobster/core/task.py#L112).

**Comment by klannon** [comment]
OK, I'll take a look and see if I can figure out what the problem is.

**Comment by klannon** [comment]
In case your following the home game, one thing I found that seems odd is that I don't see any way for `stageout exit code` to be changed from the `0` it's initialized too in `core/data/task.py.`  I'll confess that I haven't traced it through carefully enough to make sure that there isn't somewhere in a function call where it could get set.  However, it seems to me also strange that given that this job fails due to a stageout error that it doesn't set the `stageout exit code` to something other than zero.  Is that functioning as intended?

**Comment by klannon** [comment]
Since it seems that the `task exit code` is getting propagated all the way through (I see it set in `report.json`, now I'm going to start digging to try to find the logic for that decides whether a task is successful or not.  Hints welcome.

**Comment by matz-e** [comment]
Yes, I think that the `stageout exit code` got lost somewhere on the way. [Here](https://github.com/matz-e/lobster/blob/master/lobster/core/task.py#L170) we grab the exit code from WQ, we actually don't use the one in the parameter log. And [here](https://github.com/matz-e/lobster/blob/master/lobster/core/task.py#L198) we check for additional failure modes, either by resource exhaustion or by `CMSSW` failure.

**Comment by klannon** [comment]
OK, plan for the afternoon if I can squeeze it in:
1.  Add a bunch of printouts and try to run a job that fails in stageout just to verify that it's marked as success and to try to understand why.  My only hypothesis right now is that WQ has decided to give these status code 0 despite the exe having exit code != 0.
2.  I won't be able to get WQ fixed any time soon (and maybe it's not even really a bug), so I would propose that we take the hyperparanoid approach and mark the job failed if any of the following is true: task.return_status != 0, cmssw_exit_code != 0, stageout_exit_code != 0, task_exit_code !=0, and task.result != wq.WORK_QUEUE_RESULT_SUCCESS.  This will require adding some additional parsing to process_report() to pull the stageout and task exit codes from the json file.

Then I'll try to start over and see if I can get smooth sailing.  I'll probably also add in a chirp server for stageout so that we have redundant output options.

Suggestions welcome.

**Comment by klannon** [comment]
OK, I have positive confirmation that task.return_status is coming back 0 for these tasks.  I forgot to ask the wrapper to print it's return status to the log file, so I'm not sure if it's the wrapper's fault or WQ's fault.  It just occurred to me that the likely scenario is that the exit code isn't getting passed through the resource monitor so WQ is seeing that the resource monitor ended with exit code 0 instead of the wrapper's exit code...

OK, in a bit, I'll try a patch that implements the more paranoid approach above.

**Comment by matz-e** [comment]
@klannon do you have a fix for this?  Otherwise I'd hack something up quickly…

**Comment by klannon** [comment]
Sorry--I completely forgot.  I've got a fix.  Do you want it pushed 
directly onto the master or put it on a branch and make a PR?

**Comment by annawoodard** [comment]
Pushing it directly should be fine for quick fixes...

**Comment by klannon** [comment]
Darn it!  Forgot to add the "fixes" message to my commit.  This is done.  Fixed by 37c7626 directly to master.  I'm closing this issue.  I'll also ping @btovar to see if he wants to investigate any further the fact that WQ somehow was missing the non-zero return status of the jobs.

**Comment by matz-e** [comment]
Thanks!
